// leanft.sdk.sapui5/lib/list_box.js
// (c) Copyright 2015 Hewlett Packard Enterprise Development LP
/**
 * Created on 02/02/2015.
 */

var Core = require("leanft.sdk.core");
var SAPUI5BaseDesc = require("./sapui5_base.js").SAPUI5BaseDesc;
var SAPUI5BaseTestObject = require("./sapui5_base.js").SAPUI5BaseTestObject;
var HelperBehaviors = require("./helper_behaviors.js");
var WebBehaviors = require("leanft.sdk.web").Behaviors;
var SapUI5HelperFuncs = HelperBehaviors.HelperFuncs;

/**
 * Used for conversions of ListBox select type.
 * Supported ListBox select type.
 * @memberof SAPUI5
 * @enum
 */
var SelectType = {
    /**
     * A single-select list box.
     */
    singleSelection: "singleselection",

    /**
     * A combo box.
     */
    comboboxSelection: "comboboxselection",

    /**
     * An extended multi-select list.
     */
    extendedSelection: "extendedselection"
};

/**
 * @memberof sapui5
 * @ignore
 */
var SelectTypeLinker = {
    singleselection: "Single Selection",
    comboboxselection: "Combobox Select",
    extendedselection: "Extended Selection"
};

var selectTypeToNative = function(selectType){
    /*jshint eqnull:true */
    var nativeString = SelectTypeLinker[selectType.toLowerCase()];
    if (typeof(nativeString) === "undefined") {
        throw new Error("You must supply a valid selectType");
    }

    return nativeString;
};

var nativeToSelectType = function(nativeString){
    var results = Object.keys(SelectTypeLinker).filter(function(key) {
        return SelectTypeLinker[key] === nativeString;
    });

    if(results.length === 1)
        return results[0];

    throw new Error("Enum selectType name not found");
};

/**
 * Common methods for the ListBox description
 * @ignore
 * @memberof SAPUI5
 */
var ListBoxDescriptionBehavior = {
    /**
     * The selection type of the list.
     * Possible values: 'singleSelection', 'comboboxSelection', 'extendedSelection'.
     * @param {SelectType} selectType The selectType of the list box.
     * @returns {SelectType | Description} If no parameter is provided, the method acts as a getter and returns the current stored value of selectType;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    selectType: function(selectType){
        return this._propertyHandler("select type", selectType, selectTypeToNative, nativeToSelectType);
    }
};

/**
 * The description of an SAPUI5 ListBox test object.
 * @param {Object} [props] Initializes a ListBox test object using the supplied property bag.
 * @returns {ListBox} the description object that was initialized.
 * @constructor
 * @memberof SAPUI5
 * @extends SAPUI5.SAPUI5BaseDesc
 * @mixes Web.ListBoxDescriptionBehavior
 * @mixes SAPUI5.EnabledDescriptionBehavior
 * @mixes SAPUI5.ItemsCountDescriptionBehavior
 * @mixes SAPUI5.ListBoxDescriptionBehavior
 */
function ListBox(props) {
    var listBoxDes = new SAPUI5BaseDesc("ListBox");

    Core.ObjectUtils.extend(listBoxDes, WebBehaviors.ListBoxDescriptionBehavior);
    Core.ObjectUtils.extend(listBoxDes, HelperBehaviors.EnabledDescriptionBehavior);
    Core.ObjectUtils.extend(listBoxDes, HelperBehaviors.ItemsCountDescriptionBehavior);
    Core.ObjectUtils.extend(listBoxDes, ListBoxDescriptionBehavior);

    listBoxDes._defineByPropertyBag(props);

    return listBoxDes;
}

/**
 * An SAPUI5 ListBox.
 * @constructor
 * @memberof SAPUI5
 * @class ListBoxTO
 * @extends SAPUI5.SAPUI5BaseTestObject
 * @mixes Web.ListBoxTOBehavior
 * @mixes SAPUI5.EnabledTOBehavior
 * @mixes SAPUI5.MobileListBoxBehavior
 * @mixes SAPUI5.ListBoxTOBehavior
 */
function ListBoxTO(session, description) {
    var listBoxTO = new SAPUI5BaseTestObject("SAPUIList", session, description);

    Core.ObjectUtils.extend(listBoxTO, WebBehaviors.ListBoxTOBehavior);
    Core.ObjectUtils.extend(listBoxTO, HelperBehaviors.EnabledTOBehavior);
    Core.ObjectUtils.extend(listBoxTO, MobileListBoxBehavior);
    Core.ObjectUtils.extend(listBoxTO, ListBoxTOBehavior);

    return listBoxTO;
}

/**
 * @ignore
 * @memberof SAPUI5
 */
var ListBoxTOBehavior = {
    /**
     * The selection type of the list box.
     * @returns {Promise<SelectType>} a promise that is fulfilled with the selection type of the list box.
     */
    selectType: function() {
        return this._getROProperty("select type", arguments).then(nativeToSelectType);
    },

    /**
     * Indicates whether multi-selection is available for the list box.
     * @returns {Promise<boolean>} a promise that is fulfilled with an indication of whether multi-selection is available for the list box.
     */
    isMultiple: function() {
        return this._getROProperty("select type", arguments).then(function (res) {
            return res === "Extended Selection";
        });
    },

    /**
     * Returns an array of all selected items in the list.
     * @returns {Promise<string[]>} a promise that is fulfilled with an array of all selected items in the list.
     */
    selectedItems: function() {
        return this._getROProperties("selected item index", "all items").then(function(res) {
            var indicesStr = res[0];

            if (indicesStr) {
                var selectedIndices = indicesStr.split(";");
                var allItems = res[1].split(";");

                return selectedIndices.map(function(selectedIndex){
                    return allItems[selectedIndex-1];
                });
            }

            return [];
        });
    },

    /**
     * Selects a list of items in the list box.
     * @param {...number | string} item You can pass as many items as you like, each item can be either the name of the item you wish to select,
     * or the numeric index of the item in the list. Index values begin with 0.<br/>
     * Example: listBox.select("someItem", 2, "anotherItem");
     * @returns {Promise<void>} a promise that is fulfilled when the operation is performed.
     */
    select: function(item) {
        return this._executeOnItemsCollection(arguments, "Select", "ExtendSelect", Core.ArrayUtils.toArray(arguments), SapUI5HelperFuncs.itemTransform);
    },
    /**
     * Deselects a list of items in the list box.
     * @param {...number | string} item You can pass as many items as you like, each item can be either the name of the item to select,
     * or the numeric index of the item in the list. Index values begin with 0. <br/>
     * Example: listBox.deselect("someItem", 2, "anotherItem");
     * @returns {Promise<void>} a promise that is fulfilled when the operation is done.
     */
    deselect: function(item) {
        return this._executeSingleOpOnCollection(arguments, "ExtendDeselect", Core.ArrayUtils.toArray(arguments), SapUI5HelperFuncs.itemTransform);
    },

    /**
     * Enters the specified text in the list box.
     * @param {String} text The specified text to enter.
     * @returns {Promise<void>} a promise that is fulfilled after the operation is done.
     */
    set: function(text){
        return this._executeWithEvents(arguments, "Set", Core.Convert.toString(text));
    },

    /**
     * Invokes the specified filter according to the specified date.
     * @param {String} date The specified date by which to filter.
     * @returns {Promise<void>} a promise that is fulfilled after the operation is done.
     */
    filterByDate: function(date){
        return this._executeWithEvents(arguments, "FilterByDate", HelperBehaviors.DateBehavior.dateToFormatStr(date,'yyyyMMdd'));
    },

    /**
     * Enters the specified text in the list box item.
     * @param {String | Number} item The list item to enter text into. It can be the name or numeric index of the item.
     * Numeric index values begin with 0.
     * @param {String} text The specified text to enter.
     * @returns {Promise<void>} a promise that is fulfilled after the operation is done.
     */
    setItemData: function(item, text){
        return this._executeWithEvents(arguments, "SetItemData", SapUI5HelperFuncs.itemTransform(item), Core.Convert.toString(text));
    }
};

/**
 * @ignore
 * @memberof SAPUI5
 */
var MobileListBoxBehavior = {
    /**
     * Selects the inner radio button in the list item. This is an additional list box functionality on mobile devices.<br/>
     * Note: only applicable when there is an inner radio button in the list item.
     * @param item {String | Number} The list item to select. It can be the name or numeric index of the item.
     * Numeric index values begin with 0.
     * @returns a promise that is fulfilled after the operation is done.
     */
    singleSelect: function(item) {
        return this._executeWithEvents(arguments, "SingleSelect", SapUI5HelperFuncs.itemTransform(item));
    },

    /**
     * Deletes the specified item from the list. It is an additional list box functionality on mobile devices.<br/>
     * Note: only applicable when the item is deletable.
     * @param item {String | Number} The list item to delete. It can be the name or numeric index of the item.
     * Numeric index values begin with 0.
     * @returns a promise that is fulfilled after the operation is done.
     */
    delete: function(item) {
        return this._executeWithEvents(arguments, "Delete", SapUI5HelperFuncs.itemTransform(item));
    },

    /**
     * Clicks the 'More' button to expand the list and show more list items. This is an additional list box functionality on mobile devices. <br/>
     * Note: only applicable when there there is a 'More' button.
     * @returns a promise that is fulfilled after the operation is done.
     */
    more: function() {
        return this._executeWithEvents(arguments, "More");
    },

    /**
     * Selects the specified item in the list and expands the item's details. This is an additional list box functionality on mobile devices.<br/>
     * Note: only applicable when the list item has details to show.
     * @param item {String | Number} The list item. It can be the name or numeric index of the item.
     * Numeric index values begin with 0.
     */
    showItemDetails: function(item) {
        return this._executeWithEvents(arguments, "Details", SapUI5HelperFuncs.itemTransform(item));
    }
};

module.exports.ListBox = ListBox;
module.exports.ListBoxTO = ListBoxTO;
module.exports.SelectType = SelectType;
// SIG // Begin signature block
// SIG // MIIdqgYJKoZIhvcNAQcCoIIdmzCCHZcCAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // ZLgqeiTyJGyayLHkbSbDtfIogr6pDVdPXFEp7W/acj+g
// SIG // ghi2MIID7jCCA1egAwIBAgIQfpPr+3zGTlnqS5p31Ab8
// SIG // OzANBgkqhkiG9w0BAQUFADCBizELMAkGA1UEBhMCWkEx
// SIG // FTATBgNVBAgTDFdlc3Rlcm4gQ2FwZTEUMBIGA1UEBxML
// SIG // RHVyYmFudmlsbGUxDzANBgNVBAoTBlRoYXd0ZTEdMBsG
// SIG // A1UECxMUVGhhd3RlIENlcnRpZmljYXRpb24xHzAdBgNV
// SIG // BAMTFlRoYXd0ZSBUaW1lc3RhbXBpbmcgQ0EwHhcNMTIx
// SIG // MjIxMDAwMDAwWhcNMjAxMjMwMjM1OTU5WjBeMQswCQYD
// SIG // VQQGEwJVUzEdMBsGA1UEChMUU3ltYW50ZWMgQ29ycG9y
// SIG // YXRpb24xMDAuBgNVBAMTJ1N5bWFudGVjIFRpbWUgU3Rh
// SIG // bXBpbmcgU2VydmljZXMgQ0EgLSBHMjCCASIwDQYJKoZI
// SIG // hvcNAQEBBQADggEPADCCAQoCggEBALGss0lUS5ccEgrY
// SIG // JXmRIlcqb9y4JsRDc2vCvy5QWvsUwnaOQwElQ7Sh4kX0
// SIG // 6Ld7w3TMIte0lAAC903tv7S3RCRrzV9FO9FEzkMScxeC
// SIG // i2m0K8uZHqxyGyZNcR+xMd37UWECU6aq9UksBXhFpS+J
// SIG // zueZ5/6M4lc/PcaS3Er4ezPkeQr78HWIQZz/xQNRmarX
// SIG // bJ+TaYdlKYOFwmAUxMjJOxTawIHwHw103pIiq8r3+3R8
// SIG // J+b3Sht/p8OeLa6K6qbmqicWfWH3mHERvOJQoUvlXfrl
// SIG // Dqcsn6plINPYlujIfKVOSET/GeJEB5IL12iEgF1qeGRF
// SIG // zWBGflTBE3zFefHJwXECAwEAAaOB+jCB9zAdBgNVHQ4E
// SIG // FgQUX5r1blzMzHSa1N197z/b7EyALt0wMgYIKwYBBQUH
// SIG // AQEEJjAkMCIGCCsGAQUFBzABhhZodHRwOi8vb2NzcC50
// SIG // aGF3dGUuY29tMBIGA1UdEwEB/wQIMAYBAf8CAQAwPwYD
// SIG // VR0fBDgwNjA0oDKgMIYuaHR0cDovL2NybC50aGF3dGUu
// SIG // Y29tL1RoYXd0ZVRpbWVzdGFtcGluZ0NBLmNybDATBgNV
// SIG // HSUEDDAKBggrBgEFBQcDCDAOBgNVHQ8BAf8EBAMCAQYw
// SIG // KAYDVR0RBCEwH6QdMBsxGTAXBgNVBAMTEFRpbWVTdGFt
// SIG // cC0yMDQ4LTEwDQYJKoZIhvcNAQEFBQADgYEAAwmbj3nv
// SIG // f1kwqu9otfrjCR27T4IGXTdfplKfFo3qHJIJRG71betY
// SIG // fDDo+WmNI3MLEm9Hqa45EfgqsZuwGsOO61mWAK3ODE2y
// SIG // 0DGmCFwqevzieh1XTKhlGOl5QGIllm7HxzdqgyEIjkHq
// SIG // 3dlXPx13SYcqFgZepjhqIhKjURmDfrYwggSjMIIDi6AD
// SIG // AgECAhAOz/Q4yP6/NW4E2GqYGxpQMA0GCSqGSIb3DQEB
// SIG // BQUAMF4xCzAJBgNVBAYTAlVTMR0wGwYDVQQKExRTeW1h
// SIG // bnRlYyBDb3Jwb3JhdGlvbjEwMC4GA1UEAxMnU3ltYW50
// SIG // ZWMgVGltZSBTdGFtcGluZyBTZXJ2aWNlcyBDQSAtIEcy
// SIG // MB4XDTEyMTAxODAwMDAwMFoXDTIwMTIyOTIzNTk1OVow
// SIG // YjELMAkGA1UEBhMCVVMxHTAbBgNVBAoTFFN5bWFudGVj
// SIG // IENvcnBvcmF0aW9uMTQwMgYDVQQDEytTeW1hbnRlYyBU
// SIG // aW1lIFN0YW1waW5nIFNlcnZpY2VzIFNpZ25lciAtIEc0
// SIG // MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA
// SIG // omMLOUS4uyOnREm7Dv+h8GEKU5OwmNutLA9KxW7/hjxT
// SIG // VQ8VzgQ/K/2plpbZvmF5C1vJTIZ25eBDSyKV7sIrQ8Gf
// SIG // 2Gi0jkBP7oU4uRHFI/JkWPAVMm9OV6GuiKQC1yoezUvh
// SIG // 3WPVF4kyW7BemVqonShQDhfultthO0VRHc8SVguSR/yr
// SIG // rvZmPUescHLnkudfzRC5xINklBm9JYDh6NIipdC6Anqh
// SIG // d5NbZcPuF3S8QYYq3AhMjJKMkS2ed0QfaNaodHfbDlsy
// SIG // i1aLM73ZY8hJnTrFxeozC9Lxoxv0i77Zs1eLO94Ep3oi
// SIG // siSuLsdwxb5OgyYI+wu9qU+ZCOEQKHKqzQIDAQABo4IB
// SIG // VzCCAVMwDAYDVR0TAQH/BAIwADAWBgNVHSUBAf8EDDAK
// SIG // BggrBgEFBQcDCDAOBgNVHQ8BAf8EBAMCB4AwcwYIKwYB
// SIG // BQUHAQEEZzBlMCoGCCsGAQUFBzABhh5odHRwOi8vdHMt
// SIG // b2NzcC53cy5zeW1hbnRlYy5jb20wNwYIKwYBBQUHMAKG
// SIG // K2h0dHA6Ly90cy1haWEud3Muc3ltYW50ZWMuY29tL3Rz
// SIG // cy1jYS1nMi5jZXIwPAYDVR0fBDUwMzAxoC+gLYYraHR0
// SIG // cDovL3RzLWNybC53cy5zeW1hbnRlYy5jb20vdHNzLWNh
// SIG // LWcyLmNybDAoBgNVHREEITAfpB0wGzEZMBcGA1UEAxMQ
// SIG // VGltZVN0YW1wLTIwNDgtMjAdBgNVHQ4EFgQURsZpow5K
// SIG // FB7VTNpSYxc/Xja8DeYwHwYDVR0jBBgwFoAUX5r1blzM
// SIG // zHSa1N197z/b7EyALt0wDQYJKoZIhvcNAQEFBQADggEB
// SIG // AHg7tJEqAEzwj2IwN3ijhCcHbxiy3iXcoNSUA6qGTiWf
// SIG // mkADHN3O43nLIWgG2rYytG2/9CwmYzPkSWRtDebDZw73
// SIG // BaQ1bHyJFsbpst+y6d0gxnEPzZV03LZc3r03H0N45ni1
// SIG // zSgEIKOq8UvEiCmRDoDREfzdXHZuT14ORUZBbg2w6jia
// SIG // sTraCXEQ/Bx5tIB7rGn0/Zy2DBYr8X9bCT2bW+IWyhOB
// SIG // bQAuOA2oKY8s4bL0WqkBrxWcLC9JG9siu8P+eJRRw4ax
// SIG // gohd8D20UaF5Mysue7ncIAkTcetqGVvP6KUwVyyJST+5
// SIG // z3/Jvz4iaGNTmr1pdKzFHTx/kuDDvBzYBHUwggUeMIIE
// SIG // BqADAgECAhBBOdHLco8U0T0ppjKaGuk8MA0GCSqGSIb3
// SIG // DQEBCwUAMH8xCzAJBgNVBAYTAlVTMR0wGwYDVQQKExRT
// SIG // eW1hbnRlYyBDb3Jwb3JhdGlvbjEfMB0GA1UECxMWU3lt
// SIG // YW50ZWMgVHJ1c3QgTmV0d29yazEwMC4GA1UEAxMnU3lt
// SIG // YW50ZWMgQ2xhc3MgMyBTSEEyNTYgQ29kZSBTaWduaW5n
// SIG // IENBMB4XDTE2MDMyMjAwMDAwMFoXDTE3MDMyMzIzNTk1
// SIG // OVowgaQxCzAJBgNVBAYTAlVTMRMwEQYDVQQIDApDYWxp
// SIG // Zm9ybmlhMRIwEAYDVQQHDAlQYWxvIEFsdG8xKzApBgNV
// SIG // BAoMIkhld2xldHQgUGFja2FyZCBFbnRlcnByaXNlIENv
// SIG // bXBhbnkxEjAQBgNVBAsMCUNvZGUgU2lnbjErMCkGA1UE
// SIG // AwwiSGV3bGV0dCBQYWNrYXJkIEVudGVycHJpc2UgQ29t
// SIG // cGFueTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoC
// SIG // ggEBAMGgAXZcNuvF8MzEBQ9daA+60G+7667OIG2+jYQP
// SIG // NjCrcfmj+1/tNB/NRPBUAuITQCmENam4I7imHQH7Lg2O
// SIG // ilNADlkdTg2sQyvYh/vIKiLeKK1vgFOwXuqcG7Ls6xsL
// SIG // dJFhsMsq/7cSXJdWNN58wij65A0qXM3/uwcEVuayJ0XH
// SIG // KsOJoFVQD1VNBS2Y64bcAVUXvgWnmVbH/mP63s7pdjKo
// SIG // NeVOnJy8axVByDRm75M83uOQqhLhGhBxmJnWdTm1sl7L
// SIG // Drc3gZq8dPTADSUwGtkXY67vgZO4BkLxHzelYu2OMUoQ
// SIG // 8JlFDd0usxwKseyg0ZWihFUumC0kCzuv0fq1qhUCAwEA
// SIG // AaOCAW4wggFqMAkGA1UdEwQCMAAwDgYDVR0PAQH/BAQD
// SIG // AgeAMBMGA1UdJQQMMAoGCCsGAQUFBwMDMGYGA1UdIARf
// SIG // MF0wWwYLYIZIAYb4RQEHFwMwTDAjBggrBgEFBQcCARYX
// SIG // aHR0cHM6Ly9kLnN5bWNiLmNvbS9jcHMwJQYIKwYBBQUH
// SIG // AgIwGRoXaHR0cHM6Ly9kLnN5bWNiLmNvbS9ycGEwHwYD
// SIG // VR0jBBgwFoAUljtT8Hkzl699g+8uK8zKt4YecmYwKwYD
// SIG // VR0fBCQwIjAgoB6gHIYaaHR0cDovL3N2LnN5bWNiLmNv
// SIG // bS9zdi5jcmwwVwYIKwYBBQUHAQEESzBJMB8GCCsGAQUF
// SIG // BzABhhNodHRwOi8vc3Yuc3ltY2QuY29tMCYGCCsGAQUF
// SIG // BzAChhpodHRwOi8vc3Yuc3ltY2IuY29tL3N2LmNydDAR
// SIG // BglghkgBhvhCAQEEBAMCBBAwFgYKKwYBBAGCNwIBGwQI
// SIG // MAYBAQABAf8wDQYJKoZIhvcNAQELBQADggEBAIVbsb1l
// SIG // QffQZ4Ry6IkXNI1/mSIscM5U4vosrSwOou3nypo2QHpl
// SIG // C0BoUcSlavhAQ8cA5kwlxN4H+zb7bnB2lrvxCIjDT2/+
// SIG // uwWDLb1Z3GBV/Dneyr6uSKE9aUjKE7S5v7WabLcCPj1v
// SIG // 1/p6wNkPR+3czu839mJbuVOR8okTWwjZjZgs9ommE+53
// SIG // fE1cxdqzz+hD+Dukbos3+PdYaaVA2/lHoFWVlj1B9oig
// SIG // Bsv1ZbagBuEVtF1fb8uY8tDWnd4QDKE7vcq52tm9L7Xz
// SIG // B3SIyc+UTUi5BECXpqXsAHhQ9So5frKvWsELbOst/lOD
// SIG // zzMTGlOlyW6R3mc9zILt2m5Vdg4wggVZMIIEQaADAgEC
// SIG // AhA9eNf5dklgsmF99PAeyoYqMA0GCSqGSIb3DQEBCwUA
// SIG // MIHKMQswCQYDVQQGEwJVUzEXMBUGA1UEChMOVmVyaVNp
// SIG // Z24sIEluYy4xHzAdBgNVBAsTFlZlcmlTaWduIFRydXN0
// SIG // IE5ldHdvcmsxOjA4BgNVBAsTMShjKSAyMDA2IFZlcmlT
// SIG // aWduLCBJbmMuIC0gRm9yIGF1dGhvcml6ZWQgdXNlIG9u
// SIG // bHkxRTBDBgNVBAMTPFZlcmlTaWduIENsYXNzIDMgUHVi
// SIG // bGljIFByaW1hcnkgQ2VydGlmaWNhdGlvbiBBdXRob3Jp
// SIG // dHkgLSBHNTAeFw0xMzEyMTAwMDAwMDBaFw0yMzEyMDky
// SIG // MzU5NTlaMH8xCzAJBgNVBAYTAlVTMR0wGwYDVQQKExRT
// SIG // eW1hbnRlYyBDb3Jwb3JhdGlvbjEfMB0GA1UECxMWU3lt
// SIG // YW50ZWMgVHJ1c3QgTmV0d29yazEwMC4GA1UEAxMnU3lt
// SIG // YW50ZWMgQ2xhc3MgMyBTSEEyNTYgQ29kZSBTaWduaW5n
// SIG // IENBMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKC
// SIG // AQEAl4MeABavLLHSCMTXaJNRYB5x9uJHtNtYTSNiarS/
// SIG // WhtR96MNGHdou9g2qy8hUNqe8+dfJ04LwpfICXCTqdpc
// SIG // DU6kDZGgtOwUzpFyVC7Oo9tE6VIbP0E8ykrkqsDoOatT
// SIG // zCHQzM9/m+bCzFhqghXuPTbPHMWXBySO8Xu+MS09bty1
// SIG // mUKfS2GVXxxw7hd924vlYYl4x2gbrxF4GpiuxFVHU9mz
// SIG // MtahDkZAxZeSitFTp5lbhTVX0+qTYmEgCscwdyQRTWKD
// SIG // trp7aIIx7mXK3/nVjbI13Iwrb2pyXGCEnPIMlF7AVlIA
// SIG // SMzT+KV93i/XE+Q4qITVRrgThsIbnepaON2b2wIDAQAB
// SIG // o4IBgzCCAX8wLwYIKwYBBQUHAQEEIzAhMB8GCCsGAQUF
// SIG // BzABhhNodHRwOi8vczIuc3ltY2IuY29tMBIGA1UdEwEB
// SIG // /wQIMAYBAf8CAQAwbAYDVR0gBGUwYzBhBgtghkgBhvhF
// SIG // AQcXAzBSMCYGCCsGAQUFBwIBFhpodHRwOi8vd3d3LnN5
// SIG // bWF1dGguY29tL2NwczAoBggrBgEFBQcCAjAcGhpodHRw
// SIG // Oi8vd3d3LnN5bWF1dGguY29tL3JwYTAwBgNVHR8EKTAn
// SIG // MCWgI6Ahhh9odHRwOi8vczEuc3ltY2IuY29tL3BjYTMt
// SIG // ZzUuY3JsMB0GA1UdJQQWMBQGCCsGAQUFBwMCBggrBgEF
// SIG // BQcDAzAOBgNVHQ8BAf8EBAMCAQYwKQYDVR0RBCIwIKQe
// SIG // MBwxGjAYBgNVBAMTEVN5bWFudGVjUEtJLTEtNTY3MB0G
// SIG // A1UdDgQWBBSWO1PweTOXr32D7y4rzMq3hh5yZjAfBgNV
// SIG // HSMEGDAWgBR/02Wnwt3su/AwCfNDOfoCrzMxMzANBgkq
// SIG // hkiG9w0BAQsFAAOCAQEAE4UaHmmpN/egvaSvfh1hU/6d
// SIG // jF4MpnUeeBcj3f3sGgNVOftxlcdlWqeOMNJEWmHbcG/a
// SIG // IQXCLnO6SfHRk/5dyc1eA+CJnj90Htf3OIup1s+7NS8z
// SIG // WKiSVtHITTuC5nmEFvwosLFH8x2iPu6H2aZ/pFalP62E
// SIG // LinefLyoqqM9BAHqupOiDlAiKRdMh+Q6EV/WpCWJmwVr
// SIG // L7TJAUwnewusGQUioGAVP9rJ+01Mj/tyZ3f9J5THujUO
// SIG // iEn+jf0or0oSvQ2zlwXeRAwV+jYrA9zBUAHxoRFdFOXi
// SIG // vSdLVL4rhF4PpsN0BQrvl8OJIrEfd/O9zUPU8UypP7WL
// SIG // hK9k8tAUITCCBZowggOCoAMCAQICCmEZk+QAAAAAABww
// SIG // DQYJKoZIhvcNAQEFBQAwfzELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEpMCcGA1UEAxMgTWljcm9zb2Z0IENvZGUgVmVyaWZp
// SIG // Y2F0aW9uIFJvb3QwHhcNMTEwMjIyMTkyNTE3WhcNMjEw
// SIG // MjIyMTkzNTE3WjCByjELMAkGA1UEBhMCVVMxFzAVBgNV
// SIG // BAoTDlZlcmlTaWduLCBJbmMuMR8wHQYDVQQLExZWZXJp
// SIG // U2lnbiBUcnVzdCBOZXR3b3JrMTowOAYDVQQLEzEoYykg
// SIG // MjAwNiBWZXJpU2lnbiwgSW5jLiAtIEZvciBhdXRob3Jp
// SIG // emVkIHVzZSBvbmx5MUUwQwYDVQQDEzxWZXJpU2lnbiBD
// SIG // bGFzcyAzIFB1YmxpYyBQcmltYXJ5IENlcnRpZmljYXRp
// SIG // b24gQXV0aG9yaXR5IC0gRzUwggEiMA0GCSqGSIb3DQEB
// SIG // AQUAA4IBDwAwggEKAoIBAQCvJAgIKXo1nmAMqudLO07c
// SIG // fLw8RRy7K+D+KQL5VwijZIUVJ/XxrcgxiV0i6CqqpkKz
// SIG // j/i5Vbext0uz/o9+B1fs70PbZmIVYc9gDaTY3vjgw2II
// SIG // PVQT60nKWVSFJuUrjxuf6/WhkcIzSdhDY2pSS9KP6HBR
// SIG // TdGJaXvHcPaz3BJ023tdS1bTlr8Vd6Gw9KIl8q8ckmcY
// SIG // 5fQGBO+QueQA5N06tRn/Arr0PO7gi+s3i+z016zy9vA9
// SIG // r911kTMZHRxAy3QkGSGT2RT+rCpSx4/VBEnkjWNHiDxp
// SIG // g8v+R70rfk/Fla4OndTRQ8Bnc+MUCH7lP59zuDMKz10/
// SIG // NIeWiu5T6CUVAgMBAAGjgcswgcgwEQYDVR0gBAowCDAG
// SIG // BgRVHSAAMA8GA1UdEwEB/wQFMAMBAf8wCwYDVR0PBAQD
// SIG // AgGGMB0GA1UdDgQWBBR/02Wnwt3su/AwCfNDOfoCrzMx
// SIG // MzAfBgNVHSMEGDAWgBRi+wohW39DbhHaCVRQa/XSlnHx
// SIG // njBVBgNVHR8ETjBMMEqgSKBGhkRodHRwOi8vY3JsLm1p
// SIG // Y3Jvc29mdC5jb20vcGtpL2NybC9wcm9kdWN0cy9NaWNy
// SIG // b3NvZnRDb2RlVmVyaWZSb290LmNybDANBgkqhkiG9w0B
// SIG // AQUFAAOCAgEAgSqCFow0ZyvlA+s0e4yio1CK9FWG8R6M
// SIG // jq597gMZznKVGEitYhH9IP0/RwYBWuLgb4wVLE48alBs
// SIG // Czajz3oNnEK8XPgZ1WDjaebiI0FnjGiDdiuPk6MqtX++
// SIG // WfupybImj8qi84IbmD6RlSeXhmHuW10Ha82GqOJlgKji
// SIG // FeKyviMFaroM80eTTaykjAd5OcBhEjoFDYmj7J9XiYT7
// SIG // 7Mp8R2YUkdi2Dxld5rhKrLxHyHFDluYyIKXcd4b9POOL
// SIG // cdt7mwP8tx0yZOsWUqBDo/ourVmSTnzH8jNCSDhROnw4
// SIG // xxskIihAHhpGHxfbGPfwJzVsuGPZzblkXSulXu/GKbTy
// SIG // x/ghzAS6V/0BtqvGZ/nn05l/9PUi+nL1/f86HEI6ofmA
// SIG // GKXujRzUZp5FAf6q7v/7F48w9/HNKcWd7LXVSQA9hbjL
// SIG // u5M6J2pJwDCuZsn3Iygydvmkg1bISM5alqqgzAzEf7SO
// SIG // l69t41Qnw5+GwNbkcwiXBdvQVGJeA0jC1Z9/p2aM0J2w
// SIG // T9TTmF9Lesl/silS0BKAxw9Uth5nzcagbBEDhNNIdecq
// SIG // /rA7bgo6pmt2mQWj8XdoYTMURwb8U39SvZIUXEokameM
// SIG // r42QqtD2eSEbkyZ8w84evYg4kq5FxhlqSVCzBfiuWTeK
// SIG // aiUDlLFZgVDouoOAtyM19Ha5Zx1ZGK0gjZQxggRMMIIE
// SIG // SAIBATCBkzB/MQswCQYDVQQGEwJVUzEdMBsGA1UEChMU
// SIG // U3ltYW50ZWMgQ29ycG9yYXRpb24xHzAdBgNVBAsTFlN5
// SIG // bWFudGVjIFRydXN0IE5ldHdvcmsxMDAuBgNVBAMTJ1N5
// SIG // bWFudGVjIENsYXNzIDMgU0hBMjU2IENvZGUgU2lnbmlu
// SIG // ZyBDQQIQQTnRy3KPFNE9KaYymhrpPDANBglghkgBZQME
// SIG // AgEFAKB8MBAGCisGAQQBgjcCAQwxAjAAMBkGCSqGSIb3
// SIG // DQEJAzEMBgorBgEEAYI3AgEEMBwGCisGAQQBgjcCAQsx
// SIG // DjAMBgorBgEEAYI3AgEVMC8GCSqGSIb3DQEJBDEiBCCD
// SIG // MlTHKV/5HbjsX6lDiok4f9c/9TyO+tqJZNxcxkFyNjAN
// SIG // BgkqhkiG9w0BAQEFAASCAQBrj3YJ1566Y5lprp9S1JsX
// SIG // stw+0/N4GbmBpJVEDcrhwZ96i5kpBXrUyMCY+1pxRNAW
// SIG // TzFSOQyW5bHefXvzAstHc2HvHd6hI+rFJLrcm0kM1COq
// SIG // MNiq/TpNmWnFekVIN2w4Vwe1S1UjIw+E325uBSboi2JE
// SIG // o4tvcWD01dOkW5pX5228t9k7iUrH0HuMzxJ/vSLPG/Ee
// SIG // HZBB/0LRhpUvcvZc4ECcdRwqy41I8E0q8yAFsoWMVsDL
// SIG // GwveeXllr5qNfr9bMyAhfCHguNRu1iRYnmfNfsWosSen
// SIG // Ff84PgDvMKfYf9SjWqdFOK5T/rWyTt1PdQybFILuOo3E
// SIG // KC9OjtwzzKPRoYICCzCCAgcGCSqGSIb3DQEJBjGCAfgw
// SIG // ggH0AgEBMHIwXjELMAkGA1UEBhMCVVMxHTAbBgNVBAoT
// SIG // FFN5bWFudGVjIENvcnBvcmF0aW9uMTAwLgYDVQQDEydT
// SIG // eW1hbnRlYyBUaW1lIFN0YW1waW5nIFNlcnZpY2VzIENB
// SIG // IC0gRzICEA7P9DjI/r81bgTYapgbGlAwCQYFKw4DAhoF
// SIG // AKBdMBgGCSqGSIb3DQEJAzELBgkqhkiG9w0BBwEwHAYJ
// SIG // KoZIhvcNAQkFMQ8XDTE2MDUyNjE0MTgyNlowIwYJKoZI
// SIG // hvcNAQkEMRYEFO9u+S4rzpMeEwmoKQ8GSmhD3YXJMA0G
// SIG // CSqGSIb3DQEBAQUABIIBAJMNBS9OxaHm/aW08lUVW6B9
// SIG // pUmQRFWIYN//PyY/rxdkK3HHnMYT8nTZVJ3gA2sNIGrj
// SIG // m5/0ikRUmMFdKOaUwHRCqfZ2nSd2/mS6MjNiBZbejdWD
// SIG // izgsnh+qgIgxHvv+BBYbvc9Kgx0ml1LAWmCj9BhSmWzw
// SIG // le0bUS9nKC/kf1sdhPCiEa6F0Ye1Wz/F3Ccy6QqOwoMw
// SIG // CW3rXppvwBmqi8HGqnWRqzanSzkxYJWq/eYIy+sTuKxF
// SIG // jqZ+DUavHxxeQm+bAm/sVYR4ja+b6QHcoYeL+o4YdZNb
// SIG // 3GzCpM+nRFlnpee2kieove4vS+CEI7rLnEotXzqwb2CH
// SIG // nd5FdPtt3PY=
// SIG // End signature block
